Translate Chapter 16 to PTBR (#30)
JuliaIntro/JuliaIntroBR.jl: book/src/chap16.asciidoc

[[chap16]]
== Estruturas e Funções

> Agora que sabemos como criar novos tipos compostos, o próximo passo é escrever funções que recebem objetos definidos pelo programador como parâmetro e retorna-los como resultados. Neste capítulo apresentaremos um “programa funcional” e dois novos planos de desenvolvimento de programas.

Reviewer: ```suggestion
Agora que sabemos como criar novos tipos compostos, o próximo passo é escrever funções que recebem objetos definidos pelo programador como parâmetro e retornam estes objetos como resultado. Neste capítulo apresentaremos um “programa funcional” e dois novos planos de desenvolvimento de programa.
```

> Agora que sabemos como criar novos tipos compostos, o próximo passo é escrever funções que recebem objetos definidos pelo programador como parâmetro e retornam estes objetos como resultado. Neste capítulo apresentaremos um “programa funcional” e dois novos planos de desenvolvimento de programa.

Reviewer: ```suggestion
Agora que sabemos como criar novos tipos compostos, o próximo passo é escrever funções que recebem objetos definidos pelo programador como parâmetro e retornam estes objetos como resultado. Neste capítulo apresentaremos um “estilo de programação funcional” e dois novos planos de desenvolvimento de programa.
```

[[time]]

> === Hora

Reviewer: ```suggestion
=== Horário
```

> Como outro exemplo de um tipo composto, definiremos uma +struct+ chamada +minha_hora+ que grava a hora do dia. A definição de estrutura se parece como:

Reviewer: ```suggestion
Como outro exemplo de um tipo composto, definiremos uma +struct+ chamada +minha_hora+ que grava a hora do dia. A definição da estrutura se parece como:
```

> Como outro exemplo de um tipo composto, definiremos uma +struct+ chamada +minha_hora+ que grava a hora do dia. A definição da estrutura se parece como:

Reviewer: ```suggestion
Como outro exemplo de um tipo composto, definiremos uma +struct+ chamada +MeuHorário+ que guarda o horário do dia. A definição da estrutura se parece como:
```

> (((minha_hora)))((("tipo", "definido pelo programador", "minha_hora", see="minha_hora")))

Reviewer: ```suggestion
(((MeuHorário)))((("tipo", "definido pelo programador", "MeuHorário", see="MeuHorário")))
```

[source, @julia-setup chap16]
----
"""

> Representa a hora do dia.

Reviewer: ```suggestion
Representa o horário do dia.
```

campos: hora, minuto, segundo
"""

> struct minha_hora

Reviewer: ```suggestion
struct MeuHorário
```

    hora
    minuto
    segundo
end
----

> O nome +Time+ já é utilizado no Julia e para evitar um conflito de nomes, escolhemos +minha_hora+. Podemos criar um novo objeto +minha_hora_:

Reviewer: ```suggestion
O nome +Time+ já é utilizado no Julia e para evitar um conflito de nomes, escolhemos +MeuHorário+. Podemos criar um novo objeto +MeuHorário+:
```

> (((Time)))((("tipo", "Datas", "Time", see="Time")))

Reviewer: ```suggestion
(((Time)))((("tipo", "Dates", "Time", see="Time")))
```
Dates é um pacote

[source,@julia-repl-test chap16]
----

> julia> time = minha_hora(11, 59, 30)

Reviewer: ```suggestion
julia> time = MeuHorário(11, 59, 30)
```

> julia> time = MeuHorário(11, 59, 30)

Reviewer: ```suggestion
julia> horario = MeuHorário(11, 59, 30)
```

> minha_hora(11, 59, 30)

Reviewer: ```suggestion
MeuHorário(11, 59, 30)
```

----

> O diagrama de objeto para +minha_hora+ se parece com a <<fig16-1>>.

Reviewer: ```suggestion
O diagrama de objeto para +MeuHorário+ se parece com a <<fig16-1>>.
```

[[fig16-1]]
.Diagrama de objeto
image::images/fig161.svg[]

===== Exercício 16-1

> Escreva uma função chamada +imprima_hora+ que recebe um objeto +minha_hora+ e imprima-a em forma de +hora:minuto:segundo+. A instrução +@printf+ do módulo StdLib +Printf+ imprime um inteiro com o formato de sequência `"%02d"` usando pelo menos dois dígitos, incluindo um zero inicial se necessário.

Reviewer: ```suggestion
Escreva uma função chamada +imprime_hora+ que recebe um objeto +minha_hora+ e imprima-o em forma de +hora:minuto:segundo+. A instrução +@printf+ do módulo StdLib +Printf+ imprime um inteiro com o formato de sequência `"%02d"` usando pelo menos dois dígitos, incluindo um zero inicial se necessário.
```

> Escreva uma função chamada +imprime_hora+ que recebe um objeto +minha_hora+ e imprima-o em forma de +hora:minuto:segundo+. A instrução +@printf+ do módulo StdLib +Printf+ imprime um inteiro com o formato de sequência `"%02d"` usando pelo menos dois dígitos, incluindo um zero inicial se necessário.

Reviewer: ```suggestion
Escreva uma função chamada +imprime_horário+ que recebe um objeto +MeuHorário+ e imprima-o em forma de +hora:minuto:segundo+. A instrução +@printf+ do módulo StdLib +Printf+ imprime um inteiro com o formato de sequência `"%02d"` usando pelo menos dois dígitos, incluindo um zero inicial se necessário.
```

> (((imprima_hora)))((("função", "definido pelo programador", "imprima_hora", see="imprima_hora")))

Reviewer: ```suggestion
(((imprime_hora)))((("função", "definido pelo programador", "imprime_hora", see="imprime_hora")))
```

> (((imprime_hora)))((("função", "definido pelo programador", "imprime_hora", see="imprime_hora")))

Reviewer: ```suggestion
(((imprime_horário)))((("função", "definido pelo programador", "imprime_horário", see="imprime_horário")))
```

===== Exercício 16-2

> Escreva uma função booleana chamada +é_depois+ que recebe dois objetos +minha_hora+, +t1+ e +t2+, e retorna +true+ se +t1+ segue +t2+ cronologicamente e +false+ caso contrário. Desafio: não use um comando +if+.

Reviewer: ```suggestion
Escreva uma função booleana chamada +é_depois+ que recebe dois objetos +minha_hora+, +t1+ e +t2+, e retorna +true+ se +t1+ segue +t2+ cronologicamente e +false+ caso contrário. Desafio: não use uma declaração +if+.
```

> Escreva uma função booleana chamada +é_depois+ que recebe dois objetos +minha_hora+, +t1+ e +t2+, e retorna +true+ se +t1+ segue +t2+ cronologicamente e +false+ caso contrário. Desafio: não use uma declaração +if+.

Reviewer: ```suggestion
Escreva uma função booleana chamada +é_depois+ que recebe dois objetos +MeuHorário+, +t1+ e +t2+, e retorna +true+ se +t1+ segue +t2+ cronologicamente e +false+ caso contrário. Desafio: não use uma declaração +if+.
```

(((é_depois)))((("função", "definido pelo programador", "é_depois", see="é_depois")))

=== Funções Puras

> Nas próximas seções, escreveremos duas funções que adicionam valores de tempo. Elas demonstrarão dois tipos de funções: funções puras e modificadas. Elas também demonstrarão um plano de desenvolvimento que chamaremos de _protótipo e correção_, que é uma forma de combater um problema complexo começando com um protótipo simples e aos poucos lidar com as complicações.

Reviewer: ```suggestion
Nas próximas seções, escreveremos duas funções que adicionam valores de tempo. Elas demonstrarão dois tipos de funções: funções puras e modificadoras. Elas também demonstrarão um plano de desenvolvimento que chamaremos de _protótipo e correção_, que é uma forma de combater um problema complexo começando com um protótipo simples e aos poucos lidar com as complicações.
```

(((função pura)))(((protótipo e correção)))

> Aqui vai um simples protótipo de +add_tempo+:

Reviewer: ```suggestion
Aqui vai um simples protótipo de +soma_tempo+:
```

> Aqui vai um simples protótipo de +soma_tempo+:

Reviewer: ```suggestion
Aqui vai um simples protótipo de +soma_horário+:
```

> (((add_tempo)))((("função", "definido pelo programador", "add_tempo", see="add_tempo")))

Reviewer: ```suggestion
(((soma_tempo)))((("função", "definido pelo programador", "soma_tempo", see="soma_tempo")))
```

> (((soma_tempo)))((("função", "definido pelo programador", "soma_tempo", see="soma_tempo")))

Reviewer: ```suggestion
(((soma_horário)))((("função", "definido pelo programador", "soma_horário", see="soma_horário")))
```

[source,@julia-setup chap16]
----

> function add_tempo(t1, t2)

Reviewer: ```suggestion
function soma_tempo(t1, t2)
```

> function soma_tempo(t1, t2)

Reviewer: ```suggestion
function soma_horário(t1, t2)
```

>     minha_hora(t1.hora + t2.hora, t1.minuto + t2.minuto, t1.segundo + t2.segundo)

Reviewer: ```suggestion
    MeuHorário(t1.hora + t2.hora, t1.minuto + t2.minuto, t1.segundo + t2.segundo)
```

end
----

> A função cria um novo objeto +minha_hora+, inicializando os seus campos e retornando uma referência a um novo objeto. Isso é chamado de _função pura_ pois não modifica nenhum dos objetos que são passados como argumentos e não tem efeito, como exibir um valor ou obter entrada do usuário, além de retornar um valor.

Reviewer: ```suggestion
A função cria um novo objeto +minha_hora+, inicializando os seus campos e retornando uma referência ao novo objeto. Isso é chamado de _função pura_ pois não modifica nenhum dos objetos que são passados como argumentos e não tem efeito, como exibir um valor ou obter entrada do usuário, além de retornar um valor.
```

> A função cria um novo objeto +minha_hora+, inicializando os seus campos e retornando uma referência ao novo objeto. Isso é chamado de _função pura_ pois não modifica nenhum dos objetos que são passados como argumentos e não tem efeito, como exibir um valor ou obter entrada do usuário, além de retornar um valor.

Reviewer: ```suggestion
A função cria um novo objeto +MeuHorário+, inicializando os seus campos e retornando uma referência ao novo objeto. Isso é chamado de _função pura_ pois não modifica nenhum dos objetos que são passados como argumentos e não tem efeito, como exibir um valor ou obter entrada do usuário, além de retornar um valor.
```

> Para testar essa função, criaremos dois objetos +minha_hora+: +start+ que contém a hora de início de um filme, como _Monthy Python: Em Busca do Cálice Sagrado_, e +duração+ que contém o tempo de exibição do filme, que é de uma hora de 35 minutos.

Reviewer: ```suggestion
Para testar essa função, criaremos dois objetos +minha_hora+: +start+ que contém a hora de início de um filme, como _Monthy Python: Em Busca do Cálice Sagrado_, e +duração+ que contém o tempo de duração do filme, que é de uma hora e 35 minutos.
```

> Para testar essa função, criaremos dois objetos +minha_hora+: +start+ que contém a hora de início de um filme, como _Monthy Python: Em Busca do Cálice Sagrado_, e +duração+ que contém o tempo de duração do filme, que é de uma hora e 35 minutos.

Reviewer: ```suggestion
Para testar essa função, criaremos dois objetos +MeuHorário+: +início+ que contém a hora de início de um filme, como _Monthy Python: Em Busca do Cálice Sagrado_, e +duração+ que contém o tempo de duração do filme, que é de uma hora e 35 minutos.
```

> +add_tempo+ descobre quando que o filme irá terminar.

Reviewer: ```suggestion
+soma_tempo+ descobre quando o filme irá terminar.
```

> +soma_tempo+ descobre quando o filme irá terminar.

Reviewer: ```suggestion
+soma_horário+ descobre quando o filme irá terminar.
```

[source,@julia-eval chap16]
----
using Printf

> function imprima_hora(tempo::minha_hora)

Reviewer: ```suggestion
function imprima_horário(tempo::MeuHorário)
```

    @printf("%02d:%02d:%02d", tempo.hora, tempo.minuto, tempo.segundo)
end;
----

[source,@julia-repl-test chap16]
----

> julia> início = minha_hora(9, 45, 0);

Reviewer: ```suggestion
julia> início = MeuHorário(9, 45, 0);
```

> julia> duração = minha_hora(1, 35, 0);

Reviewer: ```suggestion
julia> duração = MeuHorário(1, 35, 0);
```

> julia> fim = add_tempo(início, duração);

Reviewer: ```suggestion
julia> fim = soma_horário(início, duração);
```

> julia> imprima_hora(fim)

Reviewer: ```suggestion
julia> imprima_horário(fim)
```

10:80:00
----

> O resultado, +10:80:00+ talvez não seja o que esperávamos. O problema é que essa função não lida com casos onde o número de segundos ou minutos são adicionados a mais que sessenta. Quando isso ocorre, temos que "levar" os segundos extras para a coluna de minutos ou os minutos extras para a coluna de horas. Aqui vai uma versão aprimorada:

Reviewer: ```suggestion
O resultado, +10:80:00+ talvez não seja o que esperávamos. O problema é que essa função não lida com casos no qual o número de segundos ou minutos ultrapassam sessenta. Quando isso ocorre, temos que "carregar" os segundos extras para a coluna de minutos ou os minutos extras para a coluna de horas. Aqui vai uma versão aprimorada:
```

> O resultado, +10:80:00+ talvez não seja o que esperávamos. O problema é que essa função não lida com casos no qual o número de segundos ou minutos ultrapassam sessenta. Quando isso ocorre, temos que "carregar" os segundos extras para a coluna de minutos ou os minutos extras para a coluna de horas. Aqui vai uma versão aprimorada:

Reviewer: ```suggestion
O resultado, +10:80:00+ talvez não seja o que esperávamos. O problema é que essa função não lida com casos no qual o número de segundos ou minutos ultrapassam sessenta. Quando isso ocorre, temos que “carregar” os segundos extras para a coluna de minutos ou os minutos extras para a coluna de horas. Aqui vai uma versão aprimorada:
```

[source, @julia-setup chap16]
----

> function add_tempo(t1, t2)

Reviewer: ```suggestion
function soma_tempo(t1, t2)
```

> function soma_tempo(t1, t2)

Reviewer: ```suggestion
function soma_horário(t1, t2)
```

    segundo = t1.segundo + t2.segundo
    minuto = t1.minuto + t2.minuto
    hora = t1.hora + t2.hora
    if segundo >= 60
        segundo -= 60
        minuto += 1
    end
    if minuto >= 60
        minuto -= 60
        hora += 1
    end

>     minha_hora(hora, minuto, segundo)

Reviewer: ```suggestion
    MeuHorário(hora, minuto, segundo)
```

end
----

> Apesar desta função estar correta, está começando a ficar grande. Iremos ver uma alternativa mais resumida brevemente.

Reviewer: ```suggestion
Apesar desta função estar correta, ela está começando a ficar grande. Iremos ver uma alternativa mais resumida em breve.
```

> Apesar desta função estar correta, ela está começando a ficar grande. Iremos ver uma alternativa mais resumida em breve.

Reviewer: ```suggestion
Apesar desta função estar correta, ela está começando a ficar grande. Veremos uma alternativa mais concisa em breve.
```

[[modifiers]]
=== Modificadores

> Em certos momentos, é últil para uma função modificar os objetos que recebe como parâmetros. Neste caso, as mudanção são visíveis a quem chama. Função que funcionam dessa maneiro são ditas _modificadoras_.

Reviewer: ```suggestion
Em certos momentos, é útil para uma função modificar os objetos que recebe como parâmetros. Neste caso, as mudanção são visíveis para quem a chama. Funções que funcionam dessa maneira são chamadas de _modificadoras_.
```

(((modificador)))

> +incrementa!+, que adiciona um número de segundos dados para um objeto +minha_hora+, pode ser escrito naturalmente como um modificador. Segue abaixo um rascunho:

Reviewer: ```suggestion
+incrementa!+, que adiciona um dado número de segundos à um objeto +minha_hora+, pode ser escrito naturalmente como um modificador. Aqui está um rascunho:
```

> (((incrementa)))((("função", "definido pelo programador", "incrementa!", see="incrementa!")))

Reviewer: ```suggestion
(((incrementa!)))((("função", "definido pelo programador", "incrementa!", see="incrementa!")))
```

[source,@julia-setup chap16]
----
function incrementa!(tempo, segundos)
    tempo.segundo += segundos
    if tempo.segundo >= 60
        tempo.segundo -= 60
        tempo.minuto += 1
    end
    if tempo.minuto >= 60
        tempo.minuto -= 60
        tempo.hora += 1
    end
end
----

A primeira linha executa a operação básica, o resto lida com os casos especiais que vimos anteriormente.

Essa função está correta? O que acontece se +segundos+ for muito maior que 60?

> Neste caso, não basta carregar uma vez; temos que continuar a fazer isso até que +tempo.segundo+ seja menor que 60. Uma solução é substituir os comandos +if+ com os comandos +while+. O que fará com que a função fique correta, porém, não muito eficiente.

Reviewer: ```suggestion
Neste caso, não basta carregar uma vez; temos que continuar a fazer isso até que +tempo.segundo+ seja menor que 60. Uma solução é substituir as declarações +if+ com as declarações +while+. O que fará com que a função fique correta, porém, não muito eficiente.
```

> Neste caso, não basta carregar uma vez; temos que continuar a fazer isso até que +tempo.segundo+ seja menor que 60. Uma solução é substituir as declarações +if+ com as declarações +while+. O que fará com que a função fique correta, porém, não muito eficiente.

Reviewer: ```suggestion
Neste caso, não basta carregar uma vez; temos que continuar a fazer isso até que +tempo.segundo+ seja menor que 60. Uma solução é substituir as declarações +if+ por declarações +while+. O que fará com que a função fique correta, porém, não muito eficiente.
```

===== Exercício 16-3

> Escreva uma versão correta de +incremento!+ que não conheça loop algum.

Reviewer: ```suggestion
Escreva uma versão correta de +incrementa!+ que não contenha nenhum laço.
```

> Qualquer coisa que pode ser feita com modificadores podem ser feitas com funções puras. De fato, algumas linguagens de progração permite apenas funções puras. Há algumas evidências de que programas que usam funções puras são rápidas para serem desenvolvidas e menos propenso a erros do que programas que usam modificadores. Porém, às vezes modificadores são convenientes e programas funcionais tendem a serem menos eficientes.

Reviewer: ```suggestion
Qualquer coisa que pode ser feita com modificadores também pode ser feita com funções puras. De fato, algumas linguagens de progração permitem apenas funções puras. Há algumas evidências de que programas que usam funções puras são rápidos para serem desenvolvidas e menos propensos a erros do que programas que usam modificadores. Porém, às vezes modificadores são convenientes e programas funcionais tendem a ser menos eficientes.
```

> Qualquer coisa que pode ser feita com modificadores também pode ser feita com funções puras. De fato, algumas linguagens de progração permitem apenas funções puras. Há algumas evidências de que programas que usam funções puras são rápidos para serem desenvolvidas e menos propensos a erros do que programas que usam modificadores. Porém, às vezes modificadores são convenientes e programas funcionais tendem a ser menos eficientes.

Reviewer: ```suggestion
Qualquer coisa que pode ser feita com modificadores também pode ser feita com funções puras. De fato, algumas linguagens de programação permitem apenas funções puras. Há algumas evidências de que programas que usam funções puras são mais rápidos para serem desenvolvidos e menos propensos a erros do que programas que usam modificadores. Porém, às vezes modificadores são convenientes e programas funcionais tendem a ser menos eficientes.
```

> Em geral, recomendamos que você escreva funções puras sempre que for razoável e reccorer a modificadores apenas se há uma vantagem atraente. Essa abordagem pode ser chamada de _programa funcional_.

Reviewer: ```suggestion
Em geral, recomendamos que você escreva funções puras sempre que for razoável e recorrer a modificadores apenas se há uma vantagem atraente. Essa abordagem pode ser chamada de _estilo de programação funcional_.
```

> (((programa funcional)))

Reviewer: ```suggestion
(((estilo de programação funcional)))
```

===== Exercício 16-4

> Escreva uma versão "pura" de +incremento+ que cria e retorna um novo objeto +minha_hora+ em vez de modificar o parâmetro.

Reviewer: ```suggestion
Escreva uma versão "pura" de +incrementa+ que cria e retorna um novo objeto +minha_hora+ ao invéz de modificar o parâmetro.
```

> Escreva uma versão "pura" de +incrementa+ que cria e retorna um novo objeto +minha_hora+ ao invéz de modificar o parâmetro.

Reviewer: ```suggestion
Escreva uma versão “pura” de +incrementa+ que cria e retorna um novo objeto +minha_hora+ ao invéz de modificar o parâmetro.
```

[[prototyping_versus_planning]]
=== Prototipagem Versus Planejamento

> O plano de desenvolvimento que estamos demonstrando é chamado de "protótipo e correção". Para cada função, escrevemos um protótipo que executa cálculos básicos e depois testamos, fragmentando erros no caminho.

Reviewer: ```suggestion
O plano de desenvolvimento que estamos demonstrando é chamado de “protótipo e correção”. Para cada função, escrevemos um protótipo que executava os cálculos básicos e depois testavamos-o, corrigindo os erros ao longo do caminho.
```

> (((plano de desenvolvumento de programa)))(((protótipo e correção)))

Reviewer: ```suggestion
(((plano de desenvolvimento de programa)))(((protótipo e correção)))
```

> Essa abordagem pode ser efetiva, especialmente quando você ainda não tem um entendimento profundo acerca do problema. Mas correções incrementais podem gerar códigos que por vezes são desnecessariamente complicados—deste que lide com muitos casos especiais—e ilegível—desde que seja muito difícil de saber se você encontrou todos os erros.

Reviewer: ```suggestion
Essa abordagem pode ser efetiva, especialmente quando você ainda não tem um entendimento profundo acerca do problema. Mas correções incrementais podem gerar código que é desnecessariamente complicado, já que ele lida com muitos casos especiais, e também que não é confiável, já que é difícil saber se você encontrou todos os erros.
```

> Uma alternativa é o _desenvolvimento projetado_, em que uma visão de alto nível do problema pode facilitar muito a programação. Neste caso, o insight é que um objeto Time é realmente um número de três dígitos na base 60 (consulte https://pt.wikipedia.org/wiki/Sistema_de_numera%C3%A7%C3%A3o_sexagesimal)! O atributo dos segundos é a "coluna de um", o atributo de munutos é a "coluna dos sessenta" e o atributo da hora é a "coluna dos tres mil e seiscentos".

Reviewer: ```suggestion
Uma alternativa é o _desenvolvimento projetado_, em que uma visão de alto nível do problema pode facilitar muito a programação. Neste caso, o insight é que um objeto Time é realmente um número de três dígitos na base 60 (consulte https://pt.wikipedia.org/wiki/Sistema_de_numera%C3%A7%C3%A3o_sexagesimal)! O atributo dos segundos é a “coluna de ums”, o atributo de munutos é a “coluna de sessentas” e o atributo da hora é a “coluna de três mil e seiscentos”.
```

> Uma alternativa é o _desenvolvimento projetado_, em que uma visão de alto nível do problema pode facilitar muito a programação. Neste caso, o insight é que um objeto Time é realmente um número de três dígitos na base 60 (consulte https://pt.wikipedia.org/wiki/Sistema_de_numera%C3%A7%C3%A3o_sexagesimal)! O atributo dos segundos é a “coluna de ums”, o atributo de munutos é a “coluna de sessentas” e o atributo da hora é a “coluna de três mil e seiscentos”.

Reviewer: ```suggestion
Uma alternativa é o _desenvolvimento projetado_, no qual obter uma visão de alto nível do problema pode facilitar muito a programação. Neste caso, podemos perceber que um objeto Time é na verdade um número de três dígitos na base 60 (consulte https://pt.wikipedia.org/wiki/Sistema_de_numera%C3%A7%C3%A3o_sexagesimal)! O atributo dos segundos é a “coluna de ums”, o atributo de munutos é a “coluna de sessentas” e o atributo da hora é a “coluna de três mil e seiscentos”.
```

> Uma alternativa é o _desenvolvimento projetado_, no qual obter uma visão de alto nível do problema pode facilitar muito a programação. Neste caso, podemos perceber que um objeto Time é na verdade um número de três dígitos na base 60 (consulte https://pt.wikipedia.org/wiki/Sistema_de_numera%C3%A7%C3%A3o_sexagesimal)! O atributo dos segundos é a “coluna de ums”, o atributo de munutos é a “coluna de sessentas” e o atributo da hora é a “coluna de três mil e seiscentos”.

Reviewer 2: ```suggestion
Uma alternativa é o _desenvolvimento projetado_, no qual obter uma visão de alto nível do problema pode facilitar muito a programação. Neste caso, podemos perceber que um objeto Time é na verdade um número de três dígitos na base 60 (consulte https://pt.wikipedia.org/wiki/Sistema_de_numera%C3%A7%C3%A3o_sexagesimal)! O atributo dos segundos é a “coluna de uns”, o atributo de munutos é a “coluna de sessentas” e o atributo da hora é a “coluna de três mil e seiscentos”.
```

(((desenvolvimento projetado)))

> Quando escrevemos +add_tempo+ e +incremento!+, efetivamente estamos realizando uma adição na base 60, que é justamente o porquê de termos que carregar de uma coluna para outra.

Reviewer: ```suggestion
Quando escrevemos +soma_tempo+ e +incrementa!+, efetivamente estavamos realizando uma adição na base 60, que é a razão de termos carregado de uma coluna para a próxima.
```

> Quando escrevemos +soma_tempo+ e +incrementa!+, efetivamente estavamos realizando uma adição na base 60, que é a razão de termos carregado de uma coluna para a próxima.

Reviewer: ```suggestion
Quando escrevemos +soma_horário+ e +incrementa!+, efetivamente estavamos realizando uma adição na base 60, que é a razão de termos carregado de uma coluna para a próxima.
```

> Essa observação nos sugerem uma outra abordagem de todo o problema—podemos converter objetos +minha_hora+ para inteiros e obter uma vantagem do fato de que o computador sabe como realizar artimética inteira.

Reviewer: ```suggestion
Essa observação nos sugere uma outra abordagem para todo o problema—podemos converter objetos +minha_hora+ para inteiros e obter uma vantagem do fato de que o computador sabe como realizar artimética inteira.
```

> Abaixo temos uma função que converte as horas em inteiros.

Reviewer: ```suggestion
Abaixo temos uma função que converte um objeto minha_hora para inteiros.
```

> (((conversão_inteiro)))((("função", "definido pelo programador", "conversão_inteiro", see="conversão_inteiro")))

Reviewer: ```suggestion
(((hora_para_int)))((("função", "definido pelo programador", "hora_para_int", see="hora_para_int")))
```

[source, @julia-setup chap16]
----

> function conversão_inteiro(tempo)

Reviewer: ```suggestion
function hora_para_int(tempo)
```

    minutos = tempo.hora * 60 + tempo.minuto
    segundos = minutos * 60 + tempo.segundo
end
----

> Logo baixo temos uma função que converte um inteiro a +minha_hora+ (lembre-se que +divrem+ divide o primeiro argumento pelo segundo e retorna o quociente e o resto como uma tupla):

Reviewer: ```suggestion
Aqui temos uma função que converte um inteiro para +minha_hora+ (lembre-se que +divrem+ divide o primeiro argumento pelo segundo e retorna o quociente e o resto como uma tupla):
```

> (((conversão_tempo)))((("função", "definido pelo programador", "conversão_tempo", see="conversão_tempo")))

Reviewer: ```suggestion
(((int_para_hora)))((("função", "definido pelo programador", "int_para_hora", see="int_para_hora")))
```

[source,@julia-setup chap16]
----

> function conversão_tempo()

Reviewer: ```suggestion
function int_para_hora()
```

    (minutos, segundo) = divrem(segundos, 60)
    hora, minuto = divrem(minutos, 60)
    minha_hora(hora, minuto, segundo)
end
----

> Talvez você tenha que pensar um pouco e executar alguns testes para se convercer de que essas funções estão corretas. Uma forma de testar é verificar que +conversão_inteiro(conversão_inteiro(x)) == x+ para quaisquer valores de +x+. Esse é um exemplo de consistência na verificação.

Reviewer: ```suggestion
Talvez você tenha que pensar um pouco e executar alguns testes para se convencer de que essas funções estão corretas. Uma forma de testar é verificar que +int_para_hora(hora_para_int(x)) == x+ para quaisquer valores de +x+. Esse é um exemplo de verificação de consistência.
```

> Uma vez que você esteja convicto que está correto, você poderar usar-los para reescrever +add_tempo+

Reviewer: ```suggestion
Uma vez que você esteja convencido de que elas estão corretas, você poderá usá-las para reescrever +soma_tempo+
```

> Uma vez que você esteja convencido de que elas estão corretas, você poderá usá-las para reescrever +soma_tempo+

Reviewer: ```suggestion
Uma vez que você esteja convencido de que elas estão corretas, você poderá usá-las para reescrever +soma_horário+
```

[source,@julia-setup chap16]
----

> function add_tempo(t1, t2)

Reviewer: ```suggestion
function soma_tempo(t1, t2)
```

>     segundos = conversão_inteiro(t1) + conversão_inteiro(t2)

Reviewer: ```suggestion
    segundos = tempo_para_int(t1) + tempo_para_int(t2)
```

> function soma_tempo(t1, t2)

Reviewer: ```suggestion
function soma_horário(t1, t2)
```

>     conversão_inteiro(segundos)

Reviewer: ```suggestion
    tempo_para_int(segundos)
```

>     segundos = tempo_para_int(t1) + tempo_para_int(t2)

Reviewer: ```suggestion
    segundos = hora_para_int(t1) + hora_para_int(t2)
```

>     tempo_para_int(segundos)

Reviewer: ```suggestion
    int_para_hora(segundos)
```

end
----

Essa versão é mais curta que a original e mais fácil de verificar.

===== Exercício 16-5

> Reescreva +incremento!+ usando +conversão_inteiro+ e +conversão_tempo+.

Reviewer: ```suggestion
Reescreva +incrementa!+ usando +tempo_para_int+ e +int_para_tempo+.
```

> Reescreva +incrementa!+ usando +tempo_para_int+ e +int_para_tempo+.

Reviewer: decidir entre tempo_para_int int_para_tempo e hora_para_int e int_para_hora

Author: Foi uma alteração sua. Vou ter que rever com cuidado pra ver o que, que é o que.

Reviewer: não alterei nada, se alguma coisa foi alterada foi por que eu fiz uma sugestão e você aceitou. O que tem que ser decidido é se duas funções que são chamadas de +timetoint+ e +inttotime+ traduzem pra +tempo_para_int+ e +int_para_tempo+ ou +hora/horario_para_int+ e +int_para_hora/horario+

> Algumas vezes, converter da base 60 para a base 10 e vice-versa é mais difícil que lidar com tempo. A conversão de base é mais abstracta; nossa intuição por lidar com valores de tempo é melhor.

Reviewer: ```suggestion
Algumas vezes, converter da base 60 para a base 10 e vice-versa é mais difícil do que lidar com tempo. A conversão de base é mais abstrata; a nossa intuição para lidar com valores de tempo é melhor.
```

> No entanto, se tivermos a visão de tratar horas como números de base 60 e investirmos esforço em escrever as funções de conversão (+conversão_inteiro+ e +conversão_tempo+), chegamos a um programa que é mais curto, mais fácil de ler e depurar, e mais confiável.

Reviewer: ```suggestion
Mas se nós tivermos a ideia de tratar horas como número de base 60 e investir na escrita de funções de conversão (+tempo_para_int+ e +int_para_tempo+), nós temos um programa que é menor, mais fácil de ler e depurar, e mais confiável.
```

> Também é mais fácil acrescentar recursos depois. Por exemplo, imagine subtrair dois objetos +minha_hora+ para encontrar a duração entre eles. Uma abordagem ingênua seria implementar a subtração com transporte. Porém, usar funções de conversão seria mais fácil e, provavelmente, mais correto.

Reviewer: ```suggestion
Também é mais fácil acrescentar característica depois. Por exemplo, imagine subtrair dois objetos +minha_hora+ para encontrar a duração entre eles. Uma abordagem ingênua seria implementar a subtração com empréstimo. Porém, usar funções de conversão seria mais fácil e, provavelmente, mais correto.
```

> Também é mais fácil acrescentar característica depois. Por exemplo, imagine subtrair dois objetos +minha_hora+ para encontrar a duração entre eles. Uma abordagem ingênua seria implementar a subtração com empréstimo. Porém, usar funções de conversão seria mais fácil e, provavelmente, mais correto.

Reviewer: ```suggestion
Também é mais fácil acrescentar característica depois. Por exemplo, imagine subtrair dois objetos +MeuHorário+ para encontrar a duração entre eles. Uma abordagem ingênua seria implementar a subtração com empréstimo. Porém, usar funções de conversão seria mais fácil e, provavelmente, mais correto.
```

> Ironicamente, tornar um problema mais difícil (ou mais geral) facilita (porque há menos casos especiais e menos oportunidades de erro).

Reviewer: ```suggestion
Ironicamente, as vezes tornar um problema mais difícil (ou mais geral) facilita (porque há menos casos especiais e menos oportunidades de erro).
```

[[chap16_debugging]]
=== Depurando

> Um objeto +minha_hora+ é bem formado se os valores de +minuto+ e +segundo+ estão entre 0 e 60 (incluindo 0 mas não 60) e se +hora+ é positivo. +hora+ e +minuto+ devem ser valores integrais mas talvez devessemos permitir que +segundo+ tenha uma parte fracional.

Reviewer: ```suggestion
Um objeto +MeuHorário+ é bem formado se os valores de +minuto+ e +segundo+ estão entre 0 e 60 (incluindo 0 mas não 60) e se +hora+ é positivo. +hora+ e +minuto+ devem ser valores integrais mas talvez devessemos permitir que +segundo+ tenha uma parte fracional.
```

(((depurando)))

> Requisitos como esses são ditos _invariantes_ porque sempre devem ser verdadeiros. Para dizer de outra forma, se não forem verdadeiros, algo deu errado.

Reviewer: ```suggestion
Requisitos como esses são ditos _invariantes_ porque eles sempre devem ser verdadeiros. Para dizer de outra forma, se eles não forem verdadeiros, algo deu errado.
```

(((invariante)))

> Escrever código para verificar requisitos invariantes pode ajudar a descobrir erros e encontrar suas causas. Por exemplo, você pode ter uma função como +hora_válida+, que receba um objeto +minha_hora+ e retorne +false+ se ele violar um requisito invariante:

Reviewer: ```suggestion
Escrever código para verificar requisitos invariantes pode ajudar a descobrir erros e encontrar suas causas. Por exemplo, você pode ter uma função como +hora_válida+, que receba um objeto +minha_hora+ e retorna +false+ se ela violar um requisito invariante:
```

> Escrever código para verificar requisitos invariantes pode ajudar a descobrir erros e encontrar suas causas. Por exemplo, você pode ter uma função como +hora_válida+, que receba um objeto +minha_hora+ e retorna +false+ se ela violar um requisito invariante:

Reviewer: ```suggestion
Escrever código para verificar requisitos invariantes pode ajudar a descobrir erros e encontrar suas causas. Por exemplo, você pode ter uma função como +hora_válida+, que receba um objeto +MeuHorário+ e retorna +false+ se ela violar um requisito invariante:
```

(((hora_válida)))((("função", "definido pelo programador", "hora_válida", see="hora_válida")))

[source,@julia-setup chap16]
----
function hora_válida(tempo)
    if tempo.hora < 0 || tempo.minuto < 0 || tempo.segundo < 0
        return false
    end
    if tempo.minuto >= 60 || tempo.segundo >= 60
        return false
    end
    true
end
----

No início de cada função você deve checar os argumentos para ter certeza de que eles são válidos.

> (((add_tempo)))(((error)))

Reviewer: ```suggestion
(((soma_tempo)))(((error)))
```

> (((soma_tempo)))(((error)))

Reviewer: ```suggestion
(((soma_horário)))(((error)))
```

[source,@julia-setup chap16]
----

> function add_tempo(t1,t2)

Reviewer: ```suggestion
function soma_tempo(t1,t2)
```

> function soma_tempo(t1,t2)

Reviewer: ```suggestion
function soma_horário(t1,t2)
```

    if !hora_válida(t1) || !hora_válida(t2)

>         error("objeto minha_hora inválido em add_tempo")

Reviewer: ```suggestion
        error("objeto minha_hora inválido em soma_tempo")
```

>         error("objeto minha_hora inválido em soma_tempo")

Reviewer: ```suggestion
        error("objeto MeuHorário inválido em soma_horário")
```

    end

>     segundos = conversão_inteiro(t1) + conversão_inteiro(t2)

Reviewer: ```suggestion
    segundos = hora_para_int(t1) + hora_para_int(t2)
```

>     conversão_tempo(segundos)

Reviewer: ```suggestion
   int_para_hora(segundos)
```

end
----

> Ou você pode usar uma instrução +@asset+, que verifica determinado requisito invariável e cria uma exceção se ela falhar:

Reviewer: ```suggestion
Ou você pode usar uma instrução +@asset+, que verifica determinado requisito invariável e gera uma exceção se ele falhar:
```

> Ou você pode usar uma instrução +@asset+, que verifica determinado requisito invariável e gera uma exceção se ele falhar:

Reviewer 2: ```suggestion
Ou você pode usar uma instrução +@assert+, que verifica determinado requisito invariável e gera uma exceção se ele falhar:
```

> (((@assert)))((("instrução", "Base", "@assert", see="@assert")))

Reviewer: ```suggestion
(((@assert)))((("macro", "Base", "@assert", see="@assert")))
```

[source,@julia-setup chap16]
----

> function soma_tempo(t1, t2)

Reviewer: ```suggestion
function soma_horário(t1, t2)
```

> function add_tempo(t1, t2)

Reviewer: ```suggestion
function soma_tempo(t1, t2)
```

>     @assert(hora_válida(t1) && hora_válida(t2), "objeto minha_hora inválido em add_tempo")

Reviewer: ```suggestion
    @assert(hora_válida(t1) && hora_válida(t2), "objeto MeuHorário inválido em soma_horário")
```

>     segundos = conversão_inteiro(t1) + conversão_inteiro(t2)

Reviewer: ```suggestion
    segundos =hora_para_int(t1) + hora_para_int(t2)
```

>     segundos =hora_para_int(t1) + hora_para_int(t2)

Reviewer: ```suggestion
    segundos = hora_para_int(t1) + hora_para_int(t2)
```

>     conversão_tempo(segundos)

Reviewer: ```suggestion
    int_para_tempo(segundos)
```

end
----

> instruções +@assert+ são úteis porque distinguem o código que lida com condições normais do código que verifica erros.

Reviewer: ```suggestion
Macros +@assert+ são úteis porque distinguem o código que lida com condições normais do código que verifica erros.
```

=== Glossário

protótipo e correção::
Um plano de desenvolvimento que envolve escrever um rascunho de um programa, testar e corrigir erros que são encontrados.
(((protótipo e correção)))

desenvolvimento projetado::

> Plano de desenvolvimento que implica uma compreensão de alto nível do problema e mais planejamento que desenvolvimento incremental ou desenvolvimento prototipado.

Reviewer: ```suggestion
Plano de desenvolvimento que implica uma compreensão de alto nível do problema e mais planejamento do que desenvolvimento incremental ou desenvolvimento prototipado.
```

(((desenvolvimento projetado)))

funções puras::
Função que não altera nenhum dos objetos que recebe como argumento. A maior parte das funções puras gera resultado.
(((função pura)))

modificador::
Função que modifica um ou vários dos objetos que recebe como argumento. A maior parte dos modificadores são nulos; isto é, retornam +nothing+.
(((modificador)))

programa funcional::
Um estilo de projeto de programa no qual a maioria das funções são puras.
(((programa funcional)))

invariante::
Uma condição que nunca deve mudar durante a execução de um programa.
(((invariante)))

=== Exercícios

[[ex16-1]]
===== Exercício 16-6

> Escreva uma função chamada +mulhora+ que pega um objeto +MeuHorário+ e um número, e retorna um novo objeto +MeuHorário+ que contém o produto do +MeuHorário+ original e do número.

Reviewer: ```suggestion
Escreva uma função chamada +mult_horário+ que pega um objeto +MeuHorário+ e um número, e retorna um novo objeto +MeuHorário+ que contém o produto do +MeuHorário+ original e do número.
```

> (((mulhora)))((("função", "definido pelo programador", "mulhora", see="mulhora")))

Reviewer: ```suggestion
(((mult_horário)))((("função", "definido pelo programador", "mult_horário", see="mult_horário")))
```

> Então use +multime+ para escrever uma função que receba um objeto +minha_hora+ representando o tempo até o fim de uma corrida e um número que represente a distância e retorne um objeto +minha_hora+ com o passo médio (tempo por quilômetros).

Reviewer: ```suggestion
Em seguida use +mul_hora+ para escrever uma função que receba um objeto +minha_hora+ representando o tempo até o fim de uma corrida e um número que represente a distância e retorne um objeto +minha_hora+ com o ritmo médio (tempo por quilômetro).
```

> Em seguida use +mulhora+ para escrever uma função que receba um objeto +MeuHorário+ representando o tempo até o fim de uma corrida e um número que represente a distância e retorne um objeto +MeuHorário+ com o ritmo médio (tempo por quilômetro).

Reviewer: ```suggestion
Em seguida use +mult_horário+ para escrever uma função que receba um objeto +MeuHorário+ representando o tempo até o fim de uma corrida e um número que represente a distância e retorne um objeto +MeuHorário+ com o ritmo médio (tempo por quilômetro).
```

[[ex16-2]]
===== Exercício 16-7

> O Julia fornece objetos de tempo que são similares aos objetos +minha_hora+ desse capítulo, mas eles fornecem um conjunto rico de funções e operadores. Leia a documentação em https://docs.julialang.org/en/v1/stdlib/Dates/.

Reviewer: desalinhado

Author: @joaoluisro - Apenas no Github. A numeração tá correta.

Author: @joaoluisro - Apenas no github, no código tá correto.

Reviewer 2: Falta um espaço na última linha que quebra aqui

. Escreva um programa que obtenha a data atual e imprima o dia da semana.

. Escreva um programa que aceite um aniversário como entrada e imprima a idade do usuário e o número de dias, horas, minutos e segundos até o próximo aniversário.

> . Para duas pessoas nascidas em dias diferentes, há um dia em que uma tem o dobro da idade da outra. Esse é o dia duplo deles. Escreva um programa que faça dois aniversários e calcule o dia duplo.

Reviewer: ```suggestion
. Para duas pessoas nascidas em dias diferentes, há um dia em que uma tem o dobro da idade da outra. Esse é o dia duplo deles. Escreva um programa que receba dois aniversários e calcule o dia duplo.
```

> . Para duas pessoas nascidas em dias diferentes, há um dia em que uma tem o dobro da idade da outra. Esse é o dia duplo deles. Escreva um programa que receba dois aniversários e calcule o dia duplo.

Reviewer 2: ```suggestion
. Para duas pessoas nascidas em dias diferentes, há um dia em que uma tem o dobro da idade da outra. Esse é o Dia do Dobro deles. Escreva um programa que receba dois aniversários e calcula o Dia do Dobro.
```

> . Para lhe desafiar um pouco, escreva a versão mais geral que calcula o dia em que uma pessoa está com latexmath:[\(n\)] vezes mais antiga que a outra.

Reviewer: ```suggestion
. Para lhe desafiar um pouco, escreva a versão mais geral que calcula o dia em que uma pessoa é latexmath:[\(n\)] vezes mais velha que a outra.
```

> . Para lhe desafiar um pouco, escreva a versão mais geral que calcula o dia em que uma pessoa é latexmath:[\(n\)] vezes mais velha que a outra.

Reviewer 2: ```suggestion
. Para lhe desafiar um pouco, escreva a versão mais geral que calcula o dia em que uma pessoa é latexmath:[\(n\)] vezes mais velha que a outra.

```
